### PR TITLE
cmake: Unset CXX compiler so target doesn't use host

### LINKED ISSUE #63771: Target C++ compiler is not properly selected
**Describe the bug**
`FindTargetTools.cmake` is failing in updating `CMAKE_CXX_COMPILER` as this is 
previously set in `FindHostTools.cmake`. 
The C compiler is unset, but C++ is not here
https://github.com/zephyrproject-rtos/zephyr/blob/main/cmake/modules/FindTargetTools.cmake#L93

Please also mention any information which could help others to understand
the problem you're facing:
 - What have you tried to diagnose or workaround this issue? See https://github.com/zephyrproject-rtos/zephyr/pull/63412
 - Is this a regression? no 

**To Reproduce**
1. Enable C++ configs
2. Make sure host and target compilers are different (my setup host is clang, target is gcc)
3. Compile c and c++ source.  Observe target compilers are gcc for c and clang++ for c++ source

**Expected behavior**
Target c and c++ compilers are gcc and g++

**Impact**
Fail to build target binaries with c++

**Logs and console output**
If applicable, add console logs or other types of debug information
e.g Wireshark capture or Logic analyzer capture (upload in zip archive).
copy-and-paste text and put a code fence (\`\`\`) before and after, to help
explain the issue. (if unable to obtain text log, add a screenshot)

**Environment (please complete the following information):**
 - OS: Linux
 - Toolchain Coreboot

### PR DESCRIPTION
**Pull request by asemjonovs**
CXX compiler is using host compiler when it should find target compiler
Fixes https://github.com/zephyrproject-rtos/zephyr/issues/63771

### CODE REVIEW

File: cmake/modules/FindTargetTools.cmake [hunk 1]
```diff
@@ -92,6 +92,8 @@ endif()
 # re-set it.
 unset(CMAKE_C_COMPILER)
 unset(CMAKE_C_COMPILER CACHE)
+unset(CMAKE_CXX_COMPILER)
+unset(CMAKE_CXX_COMPILER CACHE)
```
Comment (tejlmand) [inline cmake/modules/FindTargetTools.cmake:96]: A toolchain's `generic.cmake` is intended to provide a C compiler which can be used for pre-processing.

Thus i'm a little curious as to why `CMAKE_CXX_COMPILER` is set by any `generic.cmake` toolchain file.
I see such a change to LLVM generic.cmake made it in here: https://github.com/zephyrproject-rtos/zephyr/commit/fae9923ff2dedafc92dff6f79fff0c1ff50ff742
but I wonder if that itself is a mistake.

@yperess do you recall why `compiler/clang/generic.cmake` sets CMAKE_CXX_COMPILER in #52005 ?
To me that change looks wrong, so I hope you can explain the reason.

**Review by tejlmand — COMMENTED** [review]
Thanks for proposing a fix to an observed issue 👍 .

No doubt the proposed change will work, but I wonder why the CMAKE_CXX_COMPILER is set in the first place.

A toolchain's `generic.cmake` file is intended to set the C compiler for pre-processing purposes, but there should be no need to set the `CMAKE_CXX_COMPILER`.

So before a +1 on this PR, then I would like to know a little more.

- @asemjonovs Which toolchain are you using when obersving this ?
   I assume LLVM / clang.
- Can you try to comment out this line, and see if that also solves the issue:
   https://github.com/zephyrproject-rtos/zephyr/blob/7ac34d095790155d5769f7a22c0c6280fc465707/cmake/compiler/clang/generic.cmake#L8

Although i'm confident the proposed commit works, then instead of setting and clearing variables, then I prefer to not set variables in the cache in first place, if it is not needed.

So waiting for some answers before a +1.

**Comment by asemjonovs** [comment]
> Thanks for proposing a fix to an observed issue 👍 .
> 
> No doubt the proposed change will work, but I wonder why the CMAKE_CXX_COMPILER is set in the first place.
> 
> A toolchain's `generic.cmake` file is intended to set the C compiler for pre-processing purposes, but there should be no need to set the `CMAKE_CXX_COMPILER`.
> 
> So before a +1 on this PR, then I would like to know a little more.
> 
> * @asemjonovs Which toolchain are you using when obersving this ?
>   I assume LLVM / clang.
> * Can you try to comment out this line, and see if that also solves the issue:
>   https://github.com/zephyrproject-rtos/zephyr/blob/7ac34d095790155d5769f7a22c0c6280fc465707/cmake/compiler/clang/generic.cmake#L8
> 
> Although i'm confident the proposed commit works, then instead of setting and clearing variables, then I prefer to not set variables in the cache in first place, if it is not needed.
> 
> So waiting for some answers before a +1.

@tejlmand - I'm seeing this getting set to `clang++`, when I want this set to be `arm-eabi-g++`.  When I enable `cmake --tracing`  I see it getting set within our downstream implementation of generic.cmake

```
/mnt/host/source/src/third_party/zephyr/main/cmake/modules/FindHostTools.cmake(107):  include(${TOOLCHAIN_ROOT}/cmake/compiler/${COMPILER}/generic.cmake OPTIONAL )
/mnt/host/source/src/third_party/zephyr/main/cmake/modules/extensions.cmake(2684):  zephyr_var_name(${var} ${scope} expansion_var )
/mnt/host/source/src/platform/ec/zephyr/cmake/compiler/clang/generic.cmake(5):  if(EXISTS /etc/cros_chroot_version )
/mnt/host/source/src/third_party/zephyr/main/cmake/modules/extensions.cmake(2577):  if(scope STREQUAL ENV OR scope STREQUAL CACHE )
/mnt/host/source/src/platform/ec/zephyr/cmake/compiler/clang/generic.cmake(6):  set(CMAKE_C_COMPILER /usr/bin/x86_64-pc-linux-gnu-clang )
/mnt/host/source/src/third_party/zephyr/main/cmake/modules/extensions.cmake(2583):  else()
/mnt/host/source/src/platform/ec/zephyr/cmake/compiler/clang/generic.cmake(7):  set(CMAKE_CXX_COMPILER /usr/bin/x86_64-pc-linux-gnu-clang++ )
```

File: cmake/modules/FindTargetTools.cmake (hunk 1 above)
Comment (yperess) [reply cmake/modules/FindTargetTools.cmake:96]: @tejlmand I can't seem to find any notes about why I made the change in `generic.cmake`, I do seem to recall that I tried moving it around at the time and the only place that stuck was in the generic file.

Comment (yperess) [reply cmake/modules/FindTargetTools.cmake:96]: There's no hard requirement for it to be in `generic.cmake`

**Comment by tejlmand** [comment]
> @tejlmand - I'm seeing this getting set to `clang++`, when I want this set to be `arm-eabi-g++`. When I enable `cmake --tracing` I see it getting set within our downstream implementation of generic.cmake

something looks wrong in your downstream implementation of compiler support.

I notice that you're sourcing:
```
zephyr/cmake/compiler/clang/generic.cmake(5)
```
notice the `clang` there, cause that means `COMPILER` variable is set to `clang` in this line:
```
include(${TOOLCHAIN_ROOT}/cmake/compiler/${COMPILER}/generic.cmake OPTIONAL )
```
so why do you expect `arm-eabi-g++` to be set if sourcing `clang` compiler infrastructure ?
The fact that `clang/generic.cmake` is sourced indicates to me that your `ZEPHYR_TOOLCHAIN_VARIANT` is set to `llvm`.

At this point, i'm not sure anything is directly wrong in Zephyr, and I believe it's valuable to understand exactly what is going wrong in your setup.
Are you perhaps trying to switch from one toolchain to another in an existing build folder without doing a pristine build ?

That said, I think it could be wise to unset the `CMAKE_CXX_COMPILER` as a general safety pre-caution, but then the commit message should reflect that, and not because some unexplained behavior.

**Comment by asemjonovs** [comment]
@tejlmand is there anything else you'd like to see in the commit message?

**Review by jhedberg — APPROVED** [review]

**Review by fabiobaltieri — APPROVED** [review]

**Comment by tejlmand** [comment]
> @tejlmand my understanding is that Zephyr runs `FindHostTools.cmake` followed by `FindTargetTools.cmake`.

Correct, but please note that HostTools in this case is **not** necessarily the same as a host compiler.
Host tools in this case refers to tools used locally used on host for Zephyr development.
https://github.com/zephyrproject-rtos/zephyr/blob/1c2d326579ee8cea4c809eb737585537e3482677/cmake/modules/FindHostTools.cmake#L5-L6

As example a C preprocessor is needed, and Zephyr build system should still pick the pre-processor to use based on the `ZEPHYR_TOOLCHAIN_VARIANT`, as seen here:
https://github.com/zephyrproject-rtos/zephyr/blob/1c2d326579ee8cea4c809eb737585537e3482677/cmake/modules/FindHostTools.cmake#L103

A target C compiler, like a GCC based cross-compiler, can also be used as pre-processor on the host.
This is what you'll see when you set `ZEPHYR_TOOLCHAIN_VARIANT` to be a Zephyr SDK.
In this case a GCC C compiler from Zephyr SDK is used as C pre-processor, even if your standard host compiler is LLVM.

This is why I mention something looks very suspicious if `FindHostTools` includes `clang/generic.cmake` but `FindTargetTools` includes `gcc/target.cmake`, cause your `toolchain/<toolchain-name>/[generic|target].cmake` would normally not need to swap between clang and gcc.
The main reason for the `generic` and `target` split is due to gcc, as we need just a gcc C compiler for pre-processing before we know the actual target platform, and once the actual platform is know, the `[toolchain|compiler]/target.cmake` can adjust to correct cross-compiler.

But there is nothing wrong in looking up clang as host tool and use it for preprocessing, but you should not set the C++ compiler, which is also why I made this comment: https://github.com/zephyrproject-rtos/zephyr/pull/63412#discussion_r1343093841

Setting the `CMAKE_CXX_COMPILER` in generic.cmake seems to be the root cause of the error, and that should be fixed.

Now, for future reference if someone later does a git blame to understand why we clear a CMake cache variable which is not set in first place, then a clearer description as to why this is done would be nice.
As I just described, this is not supposed to happen in first place.
> FindHostTools will set the CXX compiler, when FindTargetTools runs the CXX compiler has been set to the host version.
Remember, clearing the var will result in re-invoking `find_program` on each CMake re-run, even if it could have been avoided, and thus a good description as to why we decide to do this is important.

**Review by tejlmand — CHANGES_REQUESTED** [review]
so far I prefer the root cause to be fixed, which is not setting the C++ compiler in the first place, see: #63792 

The penalty / overhead of clearing the var on each CMake re-run is that CMake will need to lookup the C++ compiler on each invocation.

Not a huge penalty, but if we decide to do so, then I need a clearer description as to why this is required.

Not to mention the principle of which other cache vars should we be clearing, just to be safe.

**Pull request closed without merging**